fs/fs_zenfs.cc

```diff
@@ -939,7 +1025,11 @@ Status ZenFS::DecodeFileDeletionFrom(Slice* input) {
     return Status::Corruption("Zone file deletion: file ID missmatch");
 
   files_.erase(fileName);
-  zoneFile.reset();
+  s = zoneFile->RemoveLinkName(fileName);
+  if (!s.ok())
+    return Status::Corruption("Zone file deletion: file links missmatch");
+
+  if (zoneFile->GetNrLinks() == 1) zoneFile.reset();
```

Comment:
Hmm, this should be == 0, right?

Reply:
No. Since we add the original filename also to linkfiles_[0], GetNrLInks() returns 1 when there are no more hardlinks to that file and can be reset.

Reply:
The existing logic before this patch was to do a zoneFile.reset() during file deletion (I think we have debated on necessity of that, but that debate is not the part of this PR). 
Now, since we have hardlinks, on file deletion we cannot just do a zoneFile.reset() , we need to do it when we reach the last  hardlink (count is one).

Reply:
So, before we had a line of code with absolutely no effect. Now we have a more complex line of code with _no effect_. It is completely redundant. It makes the code difficult to read, because nobody can figure out _why_ it is there in the first place. 

What does it do? Nothing. So remove it. It does _not_ help understand anything and it has no functional effect.

If you can tell me why we need to perform a redundant operation on a stack local smart pointer *only in the case that we have exactly on one link to a file after decoding a deletion*, then keep it. Otherwise I think it is a really bad idea.

Reply:
Well, I am aware of the things you have said, as far as the functionality goes.
I don't agree with  `nobody can figure out... ` kind of statements, because it is obvious what is happening.

If a file has say, 3 hard links, and one of them was deleted, then in the decode path, then the corresponding object's shared pointer was getting reset. Reset() in decode path was ok when each file  was an individual entity, but now there could be situations where the deleted file was a hard link(or a original file) and there are still references to the object and the decode was resetting it. So to retain earlier code as is, I added a check do the reset only if it was a last hard link.
 
I was under the impression that you wanted to remove the if condition only. But you want to remove the reset() call also, right ?  That should be ok.